File: blackjack/python-nim/partitions.nim

proc partitions*(cards: var array[0..9, int], subtotal: int): int {. exportc, dynlib .} =
  var total: int
  result = 0

  # Hit
  for i in countup(0, 9):
    if (cards[i]>0):
      cards[i] = cards[i]-1
      total = subtotal+i+1
      if (total < 21):
        # Stand
        result += 1
        # Hit again
        result += partitions(cards, total)
      elif (subtotal+i+1==21):
        # Stand; hit again is an automatic bust
        result += 1
      cards[i] = cards[i]+1
                
  return result

File: blackjack/python-zig/partitions.zig
export fn partitions(cards: &[10]i8, subtotal: u8) u32 {
  var m: u32 = 0;

  // Hit
  var i: u8 = 0;
  while (i < 10) : (i += 1) {
    if ((*cards)[i]>0) {
      var total: u8 = subtotal+i+1;
      if (total < 21) {
        // Stand
        m += 1;
        // Hit again
        (*cards)[i] -= 1;
        m += partitions(cards, total);
        (*cards)[i] += 1;
      } else if (total==21) {
        // Stand; hit again is an automatic bust
        m += 1;
      }
    }
  }
  return m;
}
